harness/determined/experimental/inference/_torch_offline_distributed_dataset.py

```diff
@@ -0,0 +1,205 @@
+        return torch.utils.data.DataLoader(
+            self._dataset, batch_sampler=batch_sampler, num_workers=self._dataloader_num_workers
+        )
+
+    def _synchronize_and_checkpoint(self, batch_idx: int):
+        """
+        Synchronize the workers and create checkpoint to record steps completed
+        """
+        if self._rank == 0:
+            self._core_context.distributed.gather(batch_idx)
+            checkpoint_metadata = {
+                "steps_completed": batch_idx + 1,
+            }
+            with self._core_context.checkpoint.store_path(checkpoint_metadata) as (path, uuid):
+                with open(os.path.join(path, "batch_completed.json"), "w") as file_obj:
+                    json.dump({"batch_completed": batch_idx}, file_obj)
+        else:
+            self._core_context.distributed.gather(batch_idx)
+
+    def _report_progress_to_master(
+        self, searcher_op: core.DummySearcherOperation, batch_idx: int
+    ) -> None:
+        completion_rate = self._calculate_progress(batch_idx)
+        searcher_op.report_progress(completion_rate)
+
+    def _calculate_progress(self, batch_idx: int):
+        records_processed = batch_idx * self._total_worker * self._batch_size
+        return records_processed / self._dataset_len
+
+    def _get_tensorboard_path(self) -> pathlib.Path:
+        """
+        Get the path where files for consumption by TensorBoard should be written
+        """
+        return self._core_context.train.get_tensorboard_path()
+
+    def set_torch_profiler(self, *args: List[str], **kwargs: Any) -> None:
+        self._torch_profiler = torch.profiler.profile(
+            on_trace_ready=torch.profiler.tensorboard_trace_handler(
+                str(self._get_tensorboard_path())
+            ),
+            *args,
+            **kwargs,
+        )
+
+    def run(self):
+        """
+        Apply per_batch_processor to the dataset
+        """
+        dataloader = self._create_dataloader()
+        batch_idx = 0
+
+        # Create dummy searcher op to report progress to master
+        dummy_searcher_op = None
+        # Initialize dummy searcher for progress report
+        if self._rank == 0:
+            dummy_searcher_op = core.DummySearcherOperation(1, True)
+
+        for idx, X in enumerate(dataloader):
+            batch_idx = idx + self._skip
+            logging.info(f"Currently processing batch {batch_idx}")
+            self._per_batch_processor.process_batch(
+                batch=X,
+                additional_info=TorchPerBatchProcessInfo(
+                    batch_idx=batch_idx + self._skip,
+                    worker_rank=self._rank,
+                    torch_profiler=self._torch_profiler,
+                ),
+            )
+
+            if (idx + 1) % self._checkpoint_interval == 0:
+                self._synchronize_and_checkpoint(batch_idx)
+
+                if self._rank == 0:
+                    self._report_progress_to_master(dummy_searcher_op, batch_idx)
```

Comment:
For progress reporting, I think it may make sense to have master report progress as a proxy, it may not be 100% accurate until we hit a sync point (interval defined by user).

If we think of progress report as an estimate, this may be ok. I do not think it is worth it to have extra sync just to ensure progress we report is 100% accurate.

Thoughts?

Reply:
yeah i think this is fine. i don't think it's worth the extra comms overhead just to report progress.